src/bls12_381/fq.rs

```diff
@@ -1120,6 +1135,32 @@ impl Fq {
         (self.0).0[5] = r11;
         self.reduce();
     }
+
+    pub fn parity(&self) -> bool {
+        (self.0).0[0] % 1 == 1
+    }
+
+    /// Hash into the field.
+    /// Hashing consists of BLAKE2b in counter mode, truncated and shaved bits,
+    /// then interpreted it (in big endian) as Fr/Fq elements.
+    pub(crate) fn hash(mut hasher:  Blake2b) -> Self {
```

Comment:
We should invoke `Blake2b` and take the output mod q, interpreted as a big-endian integer, so this has deterministic running time.